pkg/agent/nodeportlocal/nplagent_test.go

```diff
@@ -0,0 +1,354 @@
+// +build !windows
+
+// Copyright 2019 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodeportlocal
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/k8s"
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
+	npltest "github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/rules/testing"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+)
+
+type nplAnnotation struct {
+	Podport  string
+	Nodeip   string
+	Nodeport string
+}
+
+func NewPortTable(c *gomock.Controller) *portcache.PortTable {
+	ptable := portcache.PortTable{StartPort: 40000, EndPort: 45000}
+	ptable.Table = make(map[int]portcache.NodePortData)
+
+	mockTable := npltest.NewMockPodPortRules(c)
+	mockTable.EXPECT().AddRule(gomock.Any(), gomock.Any()).AnyTimes()
+	mockTable.EXPECT().DeleteRule(gomock.Any(), gomock.Any()).AnyTimes()
+
+	ptable.PodPortRules = mockTable
+	return &ptable
+}
+
+const (
+	defaultPodName  = "test-pod"
+	defaultNS       = "default"
+	defaultNodeName = "test-node"
+	defaultHostIP   = "10.10.10.10"
+	defaultPodIP    = "192.168.32.10"
+	defaultPort     = 80
+)
+
+var kubeClient *k8sfake.Clientset
+var c *k8s.Controller
+var portTable *portcache.PortTable
+
+func getTestPod() corev1.Pod {
+	testPod := corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      defaultPodName,
+			Namespace: defaultNS,
+		},
+		Spec: corev1.PodSpec{
+			NodeName: defaultNodeName,
+			Containers: []corev1.Container{
+				{
+					Ports: []corev1.ContainerPort{
+						{
+							ContainerPort: int32(defaultPort),
+						},
+					},
+				},
+			},
+		},
+		Status: corev1.PodStatus{
+			HostIP: defaultHostIP,
+			PodIP:  defaultPodIP,
+		},
+	}
+	return testPod
+}
+
+// compareWithRetry : call fn() every second and compare returned value with expected value until timeout
+// Implements a minimal functionality similar to Eventually() of github.com/onsi/gomega
+func compareWithRetry(t *testing.T, fn interface{}, expected interface{}, timeout time.Duration) {
```

Comment:
don't the utility functions in https://godoc.org/k8s.io/apimachinery/pkg/util/wait do what you want?